ldrshared/clients/messenger/base.py

```diff
@@ -0,0 +1,81 @@
+from abc import ABCMeta, abstractmethod
+from typing import Union
+
+
+class BaseMessage(metaclass=ABCMeta):
+    """
+    Generic message class
+    """
+
+    def __init__(self, message: object):
+        self.message: object = message
+
+    @abstractmethod
+    def get_attribute(self, key: str, default=None) -> str:
+        """
+        Get a single attribute as attached to the message.
+
+        Follows dictionary get pattern, i.e: .get(key, default)
+        """
+        ...
+
+    @abstractmethod
+    def get(self, key=None, default=None) -> Union[str, dict, list]:
+        """
+        Get message content. Empty parenthesis will return the whole 
+        message as dict or str (dict where castable, else str).
+
+        You can also .get(key) as per standard dictionary get
+        mechanism, which will return one of str|dict|list
+        depending on the document in question. 
+        """
+        ...
+
+
+class BaseMessenger(metaclass=ABCMeta):
+    def __init__(self, *args, **kwargs):
+        self._setup(*args, **kwargs)
+
+    @abstractmethod
+    def _setup(self, *args, **kwargs):
+        """
+        Generic setup logic to be called when the
+        class is instantiated.
+
+        Where possible take variabels from ENV VARS
+        rather than args and kwargs to avoid fouling
+        your application logic.
+        """
+        ...
+
+    @abstractmethod
+    def subscribe(self, *args, **kwargs):
+        """
+        Subscribe the messenger client to a topic
+        """
+        ...
+
+    @abstractmethod
+    def get_next_message(self) -> BaseMessage:
```

Comment:
Presumable this method blocks until the next message is returned? Is this somewhere it might be useful to use async/await?

Reply:
It's no blocking (it's built with a timeout), there's an example here: https://github.com/GSS-Cogs/linked-data-repository-shared/tree/clients/pubsub-messenger/ldrshared/clients/messenger.

roughly:
* get next unique, not processed message from buffer
* if none, fill it up message buffer for `timeout` seconds
* get next unique, not processed message from buffer
* if still none, return None.

So we're basically slurping in async futures then resolving and working through them sequentially when `get_next_message()` is called.